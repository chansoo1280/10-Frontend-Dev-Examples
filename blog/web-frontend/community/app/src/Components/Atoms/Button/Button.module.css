
.btn {
    padding: 4px 15px;
    border: 1px solid #1890FF;
    border-radius: 2px;
    display: inline-flex;
    justify-content: center;
    align-items: center;
    gap:4px;
    background: #1890FF;
    font-size:14px;
    line-height:22px;
    color:#ffffff;
}
/* .btn {
    padding: 0 10px;
    display: inline-flex;
    justify-content: center;
    align-items: center;
    min-width: 180px;
    min-height: 36px;
    border: 1px solid $white;
    border-radius: $border-radius-base;
    background: #1890FF;
    color: $primary-6;
    box-shadow: $box-shadow-sm;
    &:focus {
        outline: none;
    }
    &--danger {
        border-color: $error-color;
        color: $error-color;
    }
    &[disabled] {
        border-color: $normal-color;
        color: $normal-color;
    }
    &--hide {
        display: none;
    }
    &--icon-only {
        padding: 0;
        min-width: auto;
        width: 36px;
        height: 36px;
        border: none;
        background: transparent;
        font-size: 24px;
        &.#{$theme-name}-#{$btn-prefix-cls}--lg {
            min-width: 48px;
        }
        &.#{$theme-name}-#{$btn-prefix-cls}--default {
            background:#fbfbfb;;
            border: 1px solid #d0d9e4;
            box-shadow: 0px 3px 6px rgba(0, 0, 0, 0.16);
        }
    }
    
    &--flex {
        min-width: unset;
        flex: 1;
    }
    &--block {
        margin: 0 auto 10px;
        display: block;
        width: 100%;
        max-width: 420px;
    }
    &--fixed {
        position: fixed;
        bottom: 40px;
        right: 20px;
    }
    &--primary {
        background: $primary-6;
        border-color: $primary-6;
        color: $white;
        &.#{$theme-name}-#{$btn-prefix-cls}--danger {
            background: $error-color;
            border-color: $error-color;
            color: $white;
        }
        &[disabled] {
            background: $normal-color;
            border-color: $normal-color;
            color: $white;
        }
    }
    &--text {
        background: transparent;
        border: none;
        min-width: unset;
        padding: 0;
        box-shadow: none;
    }
    &--link {
        background: transparent;
        text-decoration: underline;
        border: none;
        min-width: unset;
        padding: 0;
        box-shadow: none;
    }
    &--round{
        border-radius: $border-radius-base;
    }
    &--circle {
        border-radius: 50%;
    }
    &--lg {
        min-height: 48px;
        font-size: 30px;
    }
    &__desc{
        word-break:keep-all;
        font-size:14px;
    }
} */