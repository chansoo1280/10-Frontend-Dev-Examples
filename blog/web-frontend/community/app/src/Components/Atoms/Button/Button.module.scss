@import "@Styles/scss/default.scss";
$btn:"btn";
.#{$btn} {
    position:relative;
    cursor: pointer;
    padding: 4px 15px;
    &--icon-only{
        padding: 4px 8px;
    }
    border: 1px solid $primary-6;
    border-radius: $border-radius;
    display: inline-flex;
    justify-content: center;
    align-items: center;
    gap:10px;
    background: $primary-6;
    font-size:14px;
    line-height:22px;
    color:$white;
    box-shadow:0 2px 0 rgba($black, 0.043);
    transition:background-color .2s ease, border-color .2s ease, color .2s ease;
    @include click-animating("");
    &:focus {
        outline: none;
    }
    &:hover, &:focus {
        @include set-color($white, $primary-5, $primary-5);
    }
    &:active {
        @include set-color($white, $primary-7, $primary-7 );
        
    }
    &[disabled] {
        @include disabled-color();
    } 
    &--secondary {
        @include set-color(rgba($black, 0.85), $white, $gray-5);
        &:hover, &:focus {
            @include set-color($primary-5, $white, $primary-5);
        }
        &[disabled] {
            @include disabled-color();
        }
    }
    &--dashed {
        @include set-color(rgba($black, 0.85), $white, $gray-5);
        border-style:dashed;
        &:hover, &:focus {
            @include set-color($primary-6, $white, $primary-6);
        }
        &[disabled] {
            @include disabled-color();
        }
    }
    &--link {
        @include set-color($primary-6, transparent, transparent);
        box-shadow: none;
        &:hover, &:focus {
            @include set-color($primary-6, transparent, transparent);
        }
        &[data-click-animating=true]:after{
            animation:none;
        }
        &[disabled] {
            @include disabled-text-color();
        }
    }
    &--text {
        @include set-color(rgba($black, 0.85), transparent, transparent);
        box-shadow: none;
        &:hover, &:focus {
            @include set-color(rgba($black, 0.85), $gray-2, transparent);
        }
        &[data-click-animating=true]:after{
            animation:none;
        }
        &[disabled] {
            @include disabled-text-color();
        }
    }


    &--danger {
        @include set-color($white, $danger-5, $danger-5 );
        &:hover, &:focus {
            @include set-color($white, $danger-4, $danger-5);
        }
        &:after{
            @include set-color("", "", rgba($danger-5, 0.1));
        }
        &[disabled] {
            @include disabled-color();
        }
        &.#{$btn}--secondary {
            @include set-color($danger-5, $white, $danger-5);
            &:hover, &:focus {
                @include set-color($danger-4, $white, $danger-4);
            }
            &[disabled] {
                @include disabled-color();
            }
        }
        &.#{$btn}--dashed {
            @include set-color($danger-5, $white, $danger-5);
            &:hover, &:focus {
                @include set-color($danger-4, $white, $danger-4);
            }
            &[disabled] {
                @include disabled-color();
            }
        }
        &.#{$btn}--link {
            @include set-color($danger-5, transparent, transparent);
            &:hover, &:focus {
                @include set-color($danger-4, transparent, transparent);
            }
            &[disabled] {
                @include disabled-text-color();
            }
        }
        &.#{$btn}--text {
            @include set-color($danger-5, transparent, transparent);
            &:hover, &:focus {
                @include set-color($danger-4, $gray-2, transparent);
            }
            &[disabled] {
                @include disabled-text-color();
            }
        }
    }
    
    &--round{
        border-radius: $border-radius;
    }
    &--circle {
        border-radius: $border-radius-circle;
    }
    &--large{
        padding:6.4px 15px;
    }
    &--small{
        padding:0px 7px;
    }
    &--hide {
        display: none;
    }
}
/* .btn {
    padding: 0 10px;
    display: inline-flex;
    justify-content: center;
    align-items: center;
    min-width: 180px;
    min-height: 36px;
    border: 1px solid $white;
    border-radius: $border-radius-base;
    background: $primary-6;
    color: $primary-6;
    box-shadow: $box-shadow-sm;
    &:focus {
        outline: none;
    }
    &--danger {
        border-color: $error-color;
        color: $error-color;
    }
    &[disabled] {
        border-color: $normal-color;
        color: $normal-color;
    }
    &--hide {
        display: none;
    }
    &--icon-only {
        padding: 0;
        min-width: auto;
        width: 36px;
        height: 36px;
        border: none;
        background: transparent;
        font-size: 24px;
        &.#{#theme-name}-#{#btn-prefix-cls}--lg {
            min-width: 48px;
        }
        &.#{#theme-name}-#{#btn-prefix-cls}--default {
            background:#fbfbfb;;
            border: 1px solid #d0d9e4;
            box-shadow: 0px 3px 6px rgba(0, 0, 0, 0.16);
        }
    }
    
    &--flex {
        min-width: unset;
        flex: 1;
    }
    &--block {
        margin: 0 auto 10px;
        display: block;
        width: 100%;
        max-width: 420px;
    }
    &--fixed {
        position: fixed;
        bottom: 40px;
        right: 20px;
    }
    &--primary {
        background: $primary-6;
        border-color: $primary-6;
        color: $white;
        &.#{#theme-name}-#{#btn-prefix-cls}--danger {
            background: $error-color;
            border-color: $error-color;
            color: $white;
        }
        &[disabled] {
            background: $normal-color;
            border-color: $normal-color;
            color: $white;
        }
    }
    &--text {
        background: transparent;
        border: none;
        min-width: unset;
        padding: 0;
        box-shadow: none;
    }
    &--link {
        background: transparent;
        text-decoration: underline;
        border: none;
        min-width: unset;
        padding: 0;
        box-shadow: none;
    }
    &--round{
        border-radius: $border-radius-base;
    }
    &--circle {
        border-radius: 50%;
    }
    &--lg {
        min-height: 48px;
        font-size: 30px;
    }
    &__desc{
        word-break:keep-all;
        font-size:14px;
    }
} */