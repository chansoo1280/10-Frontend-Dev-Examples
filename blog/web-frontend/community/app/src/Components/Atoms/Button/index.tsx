// #region Global Imports
import React from "react"
import Link from "next/link"
import classNames from "classnames"
import { UrlObject } from "url"
// #endregion Global Imports

// #region Local Imports
import styles from "./Button.module.css"
import { Icon } from "@Components"
// #endregion Local Imports

type ButtonType = "primary" | "secondary" | "dashed" | "link" | "text"
type ButtonShape = "circle" | "round"
type ButtonHTMLType = "submit" | "button" | "reset"
interface defaultProps {
    children?: React.ReactNode
    id?: string
    className?: string
    disabled?: boolean
    show?: boolean
}

interface BaseButtonProps extends defaultProps {
    icon?: React.ReactNode
    type?: ButtonType
    shape?: ButtonShape
    size?: "small" | "medium" | "large"
    danger?: boolean
    loading?: boolean
    desc?: string
}
type AnchorButtonProps = {
    href?: string | UrlObject
    target?: string
    onClick?: React.MouseEventHandler<HTMLAnchorElement>
} & BaseButtonProps &
    Omit<React.AnchorHTMLAttributes<any>, "href" | "type" | "onClick">

type NativeButtonProps = {
    htmlType?: ButtonHTMLType
    onClick?: React.MouseEventHandler<HTMLButtonElement>
} & BaseButtonProps &
    Omit<React.ButtonHTMLAttributes<any>, "type" | "onClick">

type ButtonProps = Partial<AnchorButtonProps & NativeButtonProps>

const useLoading = ({ loading = false }: Pick<ButtonProps, "loading">) => {
    const [innerLoading, setLoading] = React.useState(loading)
    React.useEffect(() => {
        let delayTimer: number | null = null
        setLoading(loading)
        return () => {
            if (delayTimer) {
                window.clearTimeout(delayTimer)
                delayTimer = null
            }
        }
    }, [loading])
    return { innerLoading }
}
const Button = (props: ButtonProps): JSX.Element => {
    const { href, icon, loading, show, htmlType = "button", size, type = "primary", shape, className, children, danger, desc, ...rest } = props
    const { innerLoading } = useLoading({ loading })

    const handleClick = (e: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement, MouseEvent>) => {
        const { onClick, disabled } = props
        if (innerLoading || disabled) {
            e.preventDefault()
            return
        }
        if (onClick === undefined) {
            return
        }
        ;(onClick as React.MouseEventHandler<HTMLButtonElement | HTMLAnchorElement>)?.(e)
    }
    const iconType = innerLoading ? "loading" : icon
    // const { name: theme } = useContext(ThemeContext)
    const prefixCls = "btn"
    const classes = classNames(
        styles[`${prefixCls}`],
        {
            [styles[`${prefixCls}--hide`]]: show === false,
            [styles[`${prefixCls}--${size}`]]: size,
            [styles[`${prefixCls}--${type}`]]: type,
            [styles[`${prefixCls}--${shape}`]]: shape,
            [styles[`${prefixCls}--icon-only`]]: !children && children !== 0 && !!iconType,
            [styles[`${prefixCls}--loading`]]: innerLoading,
            [styles[`${prefixCls}--danger`]]: danger,
        },
        className,
    )
    const iconNode = !innerLoading ? icon : <Icon iconName="loading" />
    return (
        <>
            {href !== undefined ? (
                <Link {...(rest as AnchorButtonProps)} onClick={handleClick} className={classes} href={href}>
                    {iconNode}
                    {children}
                </Link>
            ) : (
                <button {...(rest as NativeButtonProps)} onClick={handleClick} type={htmlType} className={classes}>
                    {iconNode}
                    {children}
                </button>
            )}
            {desc && <span className={styles[`${prefixCls}__desc`]}>{desc}</span>}
        </>
    )
}
export default Button
