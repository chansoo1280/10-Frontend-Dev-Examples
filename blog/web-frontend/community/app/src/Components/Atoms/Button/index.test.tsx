import { fireEvent, render, screen } from "@testing-library/react"
import Button from "./"
describe("Button", () => {
    it("renders button unchanged", () => {
        const { container } = render(<Button>클릭</Button>)
        expect(container).toMatchSnapshot()
    })
    it("should render a button with the class of primary", () => {
        render(<Button>클릭</Button>)

        const button = screen.getByRole("button")
        expect(button).toBeInTheDocument()
        expect(button).toHaveClass("btn", "btn--primary")
    })
    it("should render a disabled button with the class of primary", () => {
        render(<Button disabled>클릭</Button>)
        const button = screen.getByRole("button")
        expect(button).toBeDisabled()
    })
    it("should render a button with the class of secondary", () => {
        render(<Button type="secondary">클릭</Button>)
        const button = screen.getByRole("button")
        expect(button).toHaveClass("btn", "btn--secondary")
    })
    it("should render a disabled button with the class of secondary", () => {
        render(
            <Button disabled type="secondary">
                클릭
            </Button>,
        )
        const button = screen.getByRole("button")
        expect(button).toHaveClass("btn", "btn--secondary")
        expect(button).toBeDisabled()
    })
    it("should render a button, checks for onClick attribute and primary class", () => {
        const handleClick = jest.fn()
        render(<Button onClick={handleClick}>클릭</Button>)

        const button = screen.getByRole("button")
        fireEvent.click(button)
        expect(handleClick).toBeCalledTimes(1)
        expect(button).toHaveClass("btn", "btn--primary")
    })
    it("should render a button as a Link, checks for href attribute and primary class", () => {
        render(<Button href="/">클릭</Button>)
        const buttonAsLink = screen.getByRole("link")
        expect(buttonAsLink).toHaveAttribute("href", "/")
        expect(buttonAsLink).toHaveClass("btn", "btn--primary")
    })
})
