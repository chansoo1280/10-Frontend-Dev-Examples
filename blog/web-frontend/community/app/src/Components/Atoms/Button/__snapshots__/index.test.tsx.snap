// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders button unchanged 1`] = `
<div>
  <button
    class="btn btn--primary"
    type="button"
  >
    클릭
  </button>
</div>
`;
